src/main/java/org/opentripplanner/routing/core/RoutingContext.java

```diff
@@ -404,5 +412,13 @@ public boolean isWheelchairAccessible(Vertex v) {
     public void destroy() {
         if (origin instanceof TemporaryVertex) ((TemporaryVertex) origin).dispose();
         if (target instanceof TemporaryVertex) ((TemporaryVertex) target).dispose();
+
+        LOG.debug("ms Cleaning up {} temporary edges", temporaryEdges.size());
+        temporaryEdges.forEach(TemporaryEdge::dispose);
```

Comment:
I think:
1. We should have one way to clean up.
2. Cleanup should be enforced in a way that a developer can not fail to do so.

Witch strategy we use is not important to me, my fix was done to quick-fix the memory leak, and then to make it a little more robust. I did not spend that much time on it. But the above solution is not very robust. Consider:
```            
TemporaryPartialPatternHop newHop = shortenEnd(rr, state, reverseHop, patternArriveVertex, flagStop);
            if (newHop == null || newHop.isTrivial(rr)) {
                if (newHop != null)
                    newHop.dispose();
                continue;
            }
            rr.rctx.temporaryEdges.add(newHop);
``` 
(from the GtfsFlexGraphModifier class)
Then the above strategy fails if:
1. The developer need to remember to add the TemporaryEdge.
1. In the case of an exception after the edge is created, but before it is added to the `temporaryEdges`.
1. Also, but not as critical - the context (holding the `temporaryEdges`) need to be passed down to every place a temporary edge is created. 

The above strategy can be fixed by passing the `temporaryEdges` to all TemporaryEdge constructors, and lett the constructor add it self if it is connecting to a none temporary vertex. Another way is to store the `temporaryEdges` on the ThreadLocal object, but then you need to be sure the current execution thread is the same as the one disposing the routing context.

Reply:
Now that I think about this more, I think we should extend this strategy to all temporary edges -- ie add ALL temporary edges to a collection which is disposed at the end of a request. This would supersede #2655. I like the approach in #2655, but it won't work for GTFS-Flex, because flag stop-related edges aren't connected to the origin and destination.

Also, I created a test (somewhat contrived) which proves that it IS possible for one request to use another request's temporary edges. The only case where this will happen, again, is flag-stop edges, because other temporary edges are only accessible via temporary vertices that are created per-request, at the endpoints.

So we need a solution that satisfies the following:
- ensures all edges created by a request are disposed
- ensures that no temporary edges created by a different request are accessible from a particular request
- robust for developers, ie if a developer creates a new kind of temporary edge they don't need to remember to do 100 things

Building on @t2gran's comment, what if we do this:

1) `Graph` object has a ThreadLocal TemporaryEdges collection
2) `Edge` constructor (which all edge constructors ultimately call via `super`) checks `this instanceof TemporaryEdge`; if so, add itself to the collection
3) `RoutingContext.destroy` cleans up the temporary edges. This is ultimately called by `PlannerResource` in the finally block.
4) `State.makeState()` checks that if the edge is temporary, it's included in the TemporaryEdges collection. Otherwise, return null.

Can we guarantee that the execution thread will always be the same between 2 and 3? Are there any scenarios where 3 wouldn't happen? 

2 and 4 are awkward anti-patterns. I think the "right" solution would be for `TemporaryEdge` to be a class and not a interface, but that would also be awkward. Too bad Java doesn't allow multiple inheritance.